File: trabalho_sd/server/service/escola.go
package service

import (
	"trabalho_sd/contract"
	"trabalho_sd/dto"
	"trabalho_sd/models"
)

type Escola struct {
	BancoDeDados contract.BancoDeDadosInterface
}

func NewEscola(bancoDeDados contract.BancoDeDadosInterface) *Escola {
	return &Escola{BancoDeDados: bancoDeDados}
}

func (e *Escola) CadastrarAluno(aluno models.Aluno) (*dto.OutPutDefault, error) {
	err := e.BancoDeDados.CadastrarAluno(aluno)
	if err != nil {
		return nil, err
	}
	return &dto.OutPutDefault{
		Status:  201,
		Message: "Aluno cadastrado",
	}, nil
}

func (e *Escola) CadastrarProfessor(professor models.Professor) (*dto.OutPutDefault, error) {
	err := e.BancoDeDados.CadastrarProfessor(professor)
	if err != nil {
		return nil, err
	}
	return &dto.OutPutDefault{
		Status:  201,
		Message: "Professor cadastrado",
	}, nil

}

func (e *Escola) CadastrarDisciplina(disciplina models.Disciplina) (*dto.OutPutDefault, error) {
	err := e.BancoDeDados.CadastrarDisciplina(disciplina)
	if err != nil {
		return nil, err
	}
	return &dto.OutPutDefault{
		Status:  201,
		Message: "Disciplina cadastrado",
	}, nil
}

func (e *Escola) ListarTodasDisciplinas() (*dto.OutPutTodasDisciplinas, error) {
	data, err := e.BancoDeDados.ListarTodasDisciplinas()
	if err != nil {
		return nil, err
	}
	return &dto.OutPutTodasDisciplinas{
		Disciplinas: data,
		Status:      200,
	}, nil
}

func (e *Escola) BuscarAlunoPorCodigo(codigo string) (*dto.OutPutBuscaAlunoPorCodigo, error) {
	alunos, err := e.BancoDeDados.BuscarAlunoPorCodigo(codigo)
	if err != nil {
		return nil, err
	}

	return &dto.OutPutBuscaAlunoPorCodigo{
		Status: 200,
		Alunos: alunos,
	}, nil
}

File: trabalho_sd/client/app/proxy/escola.py
import socket
from app.models import *
from app.connection import UDPCliente

class EscolaService:
    def __init__(self,udpCliente:UDPCliente):
        self.udp = udpCliente
        

    def _processa_requisicao(self, service: str, method: str, arguments) -> Message | None:
        msg = Message(service, method, arguments.__dict__)
        msg_json = msg.to_json()
        tentativas = 3

        for _ in range(tentativas):
            try:
                self.udp.enviar_mensagem(msg_json)
                data = self.udp.receber_mensagem()
                return Message.from_json(data)
            except socket.timeout:
                print(f"Tentativa falhou, reenviando... ({_ + 1}/{tentativas})")
            except Exception as e:
                print(f"Erro ao processar requisição: {e}")
                return None
        
        return None

    def cadastrar_aluno(self, aluno: Aluno) -> Message | None:
        return self._processa_requisicao("Escola", "CadastrarAluno", aluno)

    def cadastrar_disciplina(self, disciplina: Disciplina) -> Message | None:
        return self._processa_requisicao("Escola", "CadastrarDisciplina", disciplina)

    def cadastrar_professor(self, professor: Professor) -> Message | None:
        return self._processa_requisicao("Escola", "CadastrarProfessor", professor)

    def buscar_aluno_por_codigo(self, codigo: Codigo) -> Message | None:
        return self._processa_requisicao("Escola", "BuscarAlunoPorCodigo", codigo)
    
    def listar_todas_disciplinas(self,default : Default):
        return self._processa_requisicao("Escola","ListarTodasDisciplinas",default)
        
